### PR TITLE
[mlir][ArmSME] Support vertical layout in load and store ops

### PR DESCRIPTION
**Pull request by c-rhodes**
In SME a ZA tile slice is a one-dimensional set of horizontally or vertically contiguous elements within a ZA tile. Currently the load and store ops only support horizontal tile slices. This patch adds a tile slice layout attribute to the load and store ops to support both horizontal and vertical tile slices.

When lowering from Vector dialect horizontal layout is the default.

### CODE REVIEW

File: mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp [hunk 1]
```diff
@@ -204,29 +204,51 @@ struct LoadTileSliceToArmSMELowering
...
       llvm_unreachable("unexpected element type!");
     case 8:
-      rewriter.create<arm_sme::aarch64_sme_ld1b_horiz>(loc, allActiveMask, ptr,
-                                                       tileI32, tileSliceI32);
+      if (layout == arm_sme::TileSliceLayout::Horizontal)
+        rewriter.create<arm_sme::aarch64_sme_ld1b_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+      else
+        rewriter.create<arm_sme::aarch64_sme_ld1b_vert>(loc, allActiveMask, ptr,
+                                                        tileI32, tileSliceI32);
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:219]: Wondering how to get rid of these `if` statements. You could try this:

```suggestion
      bool isHorizontal = (layout == arm_sme::TileSliceLayout::Horizonta);
      isHorizontal ? 
        rewriter.create<arm_sme::aarch64_sme_ld1b_horiz>(
            loc, allActiveMask, ptr, tileI32, tileSliceI32) :
        rewriter.create<arm_sme::aarch64_sme_ld1b_vert>(
            loc, allActiveMask, ptr, tileI32, tileSliceI32);
```

or this:

```cpp
if (layout == arm_sme::TileSliceLayout::Horizontal) {
    switch (tileElementWidth) {
    default:
       llvm_unreachable("unexpected element type!");
    case 8:
    // (...)
    break;
} else {
    switch (tileElementWidth) {
    default:
       llvm_unreachable("unexpected element type!");
    case 8:
    // (...)
    break;
}
}
```

File: mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt [hunk 2]
```diff
@@ -4,3 +4,10 @@ add_mlir_doc(ArmSME ArmSME Dialects/ -gen-dialect-doc -dialect=arm_sme)
 set(LLVM_TARGET_DEFINITIONS ArmSME.td)
 mlir_tablegen(ArmSMEConversions.inc -gen-llvmir-conversions)
 add_public_tablegen_target(MLIRArmSMEConversionsIncGen)
+
+mlir_tablegen(ArmSMEEnums.h.inc -gen-enum-decls)
+mlir_tablegen(ArmSMEEnums.cpp.inc -gen-enum-defs)
+mlir_tablegen(ArmSMEAttrDefs.h.inc -gen-attrdef-decls -attrdefs-dialect=arm_sme)
+mlir_tablegen(ArmSMEAttrDefs.cpp.inc -gen-attrdef-defs -attrdefs-dialect=arm_sme)
+add_public_tablegen_target(MLIRArmSMEEnumsIncGen)
+add_dependencies(mlir-headers MLIRArmSMEEnumsIncGen)
```
Comment (banach-space) [inline mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt:13]: Why do we need all these changes? (there's quite a lot in here)

File: mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  // Load tile from "mem1" vertically.
+  %0 = arm_sme.tile_load <ver>, %mem1[%c0, %c0] : memref<?xi32>, vector<[4]x[4]xi32>
+
+  // Store tile back to "mem2" to print.
+  // TODO: Support vector.print for 2-D scalable vectors so don't have to spill
+  // to memory and reload to print.
+  vector.store %0, %mem2[%c0] : memref<?xi32>, vector<[4]x[4]xi32>
+
+  // Dump "mem1". The smallest SVL is 128-bits so the tile will be at least
+  // 4x4xi32.
```
Comment (banach-space) [inline mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir:75]: I love this test - it does a great job illustrating the concept of horizontal vs vertical loads!

[nit] Worth adding a note that this is the original layout, e.g.:


```suggestion
  // 1. ORIGINAL LAYOUT
  // Dump "mem1". The smallest SVL is 128-bits so the tile will be at least
  // 4x4xi32.
```

File: mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  // CHECK:      TILE END
+  func.call @printTileBegin() : () -> ()
+  scf.for %i = %c0 to %za_s_size step %svl_s {
+    %tileslice = vector.load %mem1[%i] : memref<?xi32>, vector<[4]xi32>
+    vector.print %tileslice : vector<[4]xi32>
+  }
+  func.call @printTileEnd() : () -> ()
+
+  // Dump "mem2". The smallest SVL is 128-bits so the tile will be at least
+  // 4x4xi32.
```
Comment (banach-space) [inline mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir:91]:
```suggestion
  // 2. VERTICAL LAYOUT
  // Dump "mem2". The smallest SVL is 128-bits so the tile will be at least
  // 4x4xi32.
```

File: mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir [hunk 5]
```diff
@@ -0,0 +1,401 @@
+// RUN: mlir-opt %s -convert-vector-to-llvm="enable-arm-sme" -cse -canonicalize -split-input-file -verify-diagnostics | FileCheck %s
+
+// Test conversion of higher-level ArmSME ops to LLVM intrinsics.
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir:3]: By "higher-level" you basically mean the custom ops (as opposed the intrinsics), right? As in, there are 2 categories ATM, correct?

File: mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+func.func @arm_sme_load_tile_slice_hor_i8(%src : memref<?x?xi8>, %tile : vector<[16]x[16]xi8>, %tile_slice_index : index) {
+  %c0 = arith.constant 0 : index
+  %tile_update = arm_sme.load_tile_slice <hor>, %src[%c0], %tile, %tile_slice_index : memref<?x?xi8>, vector<[16]x[16]xi8>
+  return
+}
+
+// -----
+
+// CHECK-LABEL: @arm_sme_load_tile_slice_hor_i16
+// CHECK: arm_sme.intr.ld1h.horiz
```
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir:36]: I'd be tempted to check the types as well:
```suggestion
// CHECK: arm_sme.intr.ld1h.horiz (%{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}) : (vector<[8]xi1>, !llvm.ptr, i32, i32) -> ()
```

File: mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt (hunk 2 above)
Comment (c-rhodes) [reply mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt]: this is for the attribute, see: https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/#cmake-targets

It seems just `add_public_tablegen_target(MLIRArmSMEAttrDefsIncGen)` works tho so replaced bottom two lines with that.

File: mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp (hunk 1 above)
Comment (c-rhodes) [reply mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp]: I've implemented the latter suggestion

File: mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir (hunk 5 above)
Comment (c-rhodes) [reply mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir:3]: Not really sure how I can make this any clearer than it is

File: mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir (hunk 6 above)
Comment (c-rhodes) [reply mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir]: Done

File: mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir (hunk 4 above)
Comment (c-rhodes) [reply mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir:93]: Done

File: mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir (hunk 3 above)
Comment (c-rhodes) [reply mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir:76]: Done

File: mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir (hunk 5 above)
Comment (banach-space) [reply mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir:3]: Perhaps just drop "higher-level"? That's the bit that confused me.

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

File: mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir (hunk 5 above)
Comment (c-rhodes) [reply mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir]: Done

**Review by MacDue — APPROVED** [review]

File: mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp [hunk 7]
```diff
@@ -204,30 +204,59 @@ struct LoadTileSliceToArmSMELowering
...
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
+      case 64:
+        rewriter.create<arm_sme::aarch64_sme_ld1d_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
+      case 128:
+        rewriter.create<arm_sme::aarch64_sme_ld1q_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
```
Comment (dcaballe) [inline mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:233]: worth refactor this into a template function?

File: mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp [hunk 8]
```diff
@@ -280,29 +309,58 @@ struct StoreTileSliceToArmSMELowering
...
+            storeTileSliceOp, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
+      case 64:
+        rewriter.replaceOpWithNewOp<arm_sme::aarch64_sme_st1d_horiz>(
+            storeTileSliceOp, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
+      case 128:
+        rewriter.replaceOpWithNewOp<arm_sme::aarch64_sme_st1q_horiz>(
+            storeTileSliceOp, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
```
Comment (dcaballe) [inline mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:337]: same?

**Review by dcaballe — APPROVED** [review]
Awesome! Just a small refactoring comment

File: mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp (hunk 7 above)
Comment (c-rhodes) [reply mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:233]: it's not obvious to me how I could improve this with a template function, could you give us some pointers?

**Comment by c-rhodes** [comment]
thanks for reviewing all. I've updated this to make tile slice layout optional with a default of horizontal, as suggested by @dcaballe on #66760. Optional custom attributes aren't supported as the first argument in the asm string so I've moved layout to the end of the asm string for both loads and stores for consistency. Quite a lot of churn to support this unfortunately, would appreciate a second pair of eyes on the changes before I land.

**Comment by c-rhodes** [comment]
Forgot to mention, also changes the format in the IR from:
`<hor> -> <horizontal>`
`<ver> -> <vertical>`
as suggested on https://github.com/llvm/llvm-project/pull/66760.

File: mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp (hunk 7 above)
Comment (banach-space) [reply mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:233]: This is the only thing that came to my mind:
```cpp
template <int N>
void callLoadIntrinsic(ConversionPatternRewriter &rewriter,
                       arm_sme::StoreTileSliceOp storeTileSliceOp,
                       mlir::vector::SplatOp allActiveMask, Value ptr, Value tileI32,
                       mlir::arith::IndexCastUIOp tileSliceI32) {
}

template <>
void callLoadIntrinsic<8>(ConversionPatternRewriter &rewriter,
                          arm_sme::StoreTileSliceOp storeTileSliceOp,
                          mlir::vector::SplatOp allActiveMask, Value ptr,
                          Value tileI32,
                          mlir::arith::IndexCastUIOp tileSliceI32) {
  rewriter.replaceOpWithNewOp<arm_sme::aarch64_sme_st1b_horiz>(
      storeTileSliceOp, allActiveMask, ptr, tileI32, tileSliceI32);
}
```

But that wouldn't be an improvement IMHO, so I'd go ahead with what you have here already.

Comment (dcaballe) [reply mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:233]: You could have a template function with a single switch and then:
`callLoadIntrinsic<arm_sme::aarch64_sme_ld1b_horiz, arm_sme::aarch64_sme_ld1h_horiz...>(...)`
`callLoadIntrinsic<the vertical ones>(...)`

That should reduce the code by 2x but maybe I'm missing something :)

**Pull request merged**